File: packages/android-im-ui/src/test/kotlin/com/flare/im/ui/ImageGalleryTest.kt
package com.flare.im.ui

import java.io.File
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertIs
import kotlin.test.assertNotNull
import kotlin.test.assertNull
import kotlin.test.assertTrue

/** The shared gallery table (`spec/image-gallery-vectors.json`) and what a tap on a picture presents. */
class ImageGalleryTest {
    private fun vectorsFile(): File {
        var dir: File? = File(".").absoluteFile
        while (dir != null) {
            val candidate = File(dir, "spec/image-gallery-vectors.json")
            if (candidate.isFile) return candidate
            dir = dir.parentFile
        }
        error("spec/image-gallery-vectors.json not found")
    }

    private fun message(raw: Map<*, *>): FlareMessageData {
        val refs = (raw["images"] as? List<*>).orEmpty().map { it as String }
        val content = when (raw["kind"]) {
            "image" -> FlareImageContent(refs.first())
            "imageGroup" -> FlareImageGroupContent(refs.map { FlareImageContent(it) })
            "sticker" -> FlareStickerContent("https://cdn.example/s.webp")
            "video" -> FlareVideoContent("https://cdn.example/v.mp4")
            else -> FlareTextContent("明天见")
        }
        val lifecycle = if (raw["recalled"] == true) FlareMessageLifecycle(mutation = FlareMessageMutationState.Recalled) else null
        return FlareMessageData(id = raw["id"] as String, senderId = "u", senderName = "U", content = content, lifecycle = lifecycle)
    }

    @Test fun everyCaseBuildsTheGalleryTheTableSays() {
        val cases = (FlareJson.parse(vectorsFile().readText()) as Map<*, *>)["cases"] as List<*>
        assertTrue(cases.size >= 4)
        for (raw in cases) {
            val case = raw as Map<*, *>
            val id = case["id"] as String
            val items = flareImageGalleryItems((case["messages"] as List<*>).map { message(it as Map<*, *>) })
            assertEquals(case["items"], items.map { "${it.messageId}#${it.index}" }, id)
            for (rawOpen in case["opens"] as List<*>) {
                val open = rawOpen as Map<*, *>
                val start = flareImageGalleryStart(items, open["message"] as String, (open["index"] as Double).toInt())
                assertEquals((open["start"] as Double?)?.toInt(), start, "$id ${open["message"]}#${open["index"]}")
            }
        }
    }

    private val timeline = listOf(
        FlareMessageData("m1", "u", "U", FlareImageContent("https://cdn/1.jpg")),
        FlareMessageData("m2", "u", "U", FlareImageGroupContent(listOf(FlareImageContent("https://cdn/2.jpg"), FlareImageContent("", thumbnailUrl = "https://cdn/3t.jpg")))),
    )

    @Test fun aTapInATimelineOpensTheGalleryElseThePictureAlone() {
        val gallery = FlareTimelineGallery(flareImageGalleryItems(timeline), download = null)
        assertEquals(
            FlareMediaPresentation.Gallery(listOf("https://cdn/1.jpg", "https://cdn/2.jpg", "https://cdn/3t.jpg"), 2),
            flareImagePresentation(gallery, "m2", 1, "https://cdn/3t.jpg"),
        )
        assertEquals(FlareMediaPresentation.Image("https://cdn/x.jpg"), flareImagePresentation(null, "m2", 1, "https://cdn/x.jpg"))
        assertEquals(FlareMediaPresentation.Image("https://cdn/x.jpg"), flareImagePresentation(gallery, "m9", 0, "https://cdn/x.jpg"))
        assertEquals(FlareMediaPresentation.Image("https://cdn/x.jpg"), flareImagePresentation(gallery, null, 0, "https://cdn/x.jpg"))
    }

    /**
     * The download key follows what is on screen: a gallery page downloads its own picture with the message it belongs
     * to — never the tapped body's picture — and a picture alone keeps the key its body was given.
     */
    @Test fun aGalleryPageDownloadsItsOwnPictureAndAPictureAloneItsBodysKey() {
        val saved = mutableListOf<String>()
        val gallery = FlareTimelineGallery(flareImageGalleryItems(timeline)) { saved += "${it.messageId}#${it.index}" }
        val opened = flareImagePresentation(gallery, "m2", 1, "https://cdn/3t.jpg", onDownload = { saved += "body" })
        val pages = assertIs<FlareMediaPresentation.Gallery>(opened).onDownload
        assertNotNull(pages)
        pages(2)
        pages(0)
        assertEquals(listOf("m2#1", "m1#0"), saved)

        val silent = flareImagePresentation(FlareTimelineGallery(gallery.items, download = null), "m2", 1, "https://cdn/3t.jpg", onDownload = { saved += "body" })
        assertNull(assertIs<FlareMediaPresentation.Gallery>(silent).onDownload)

        saved.clear()
        val alone = assertIs<FlareMediaPresentation.Image>(flareImagePresentation(gallery, "m9", 0, "https://cdn/x.jpg", onDownload = { saved += "body" }))
        assertNotNull(alone.onDownload).invoke()
        assertEquals(listOf("body"), saved)
        assertNull(assertIs<FlareMediaPresentation.Image>(flareImagePresentation(null, "m2", 1, "https://cdn/x.jpg")).onDownload)
    }
}

File: packages/flutter-im-ui/test/image_gallery_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:flare_im_ui/flare_im_ui.dart';
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

/// The shared gallery table (`spec/image-gallery-vectors.json`) and the paging
/// preview a timeline opens.
void main() {
  final table =
      jsonDecode(
            File('../../spec/image-gallery-vectors.json').readAsStringSync(),
          )
          as Map<String, dynamic>;
  final cases = (table['cases'] as List).cast<Map<String, dynamic>>();

  FlareMessageData message(Map<String, dynamic> raw) {
    final refs = (raw['images'] as List? ?? const []).cast<String>();
    final content = switch (raw['kind']) {
      'image' => FlareImageContent(url: refs.first),
      'imageGroup' => FlareImageGroupContent(
        images: [for (final ref in refs) FlareImageContent(url: ref)],
      ),
      'sticker' => const FlareStickerContent(url: 'https://cdn.example/s.webp'),
      'video' => const FlareVideoContent(url: 'https://cdn.example/v.mp4'),
      _ => const FlareTextContent('明天见'),
    };
    return FlareMessageData(
      id: raw['id'] as String,
      senderId: 'u',
      senderName: 'U',
      content: content,
      lifecycle: raw['recalled'] == true
          ? const FlareMessageLifecycle(
              mutation: FlareMessageMutationState.recalled,
            )
          : null,
    );
  }

  group('a conversation\'s image gallery', () {
    test('the shared table is whole', () {
      expect(cases.length, greaterThanOrEqualTo(4));
    });
    for (final c in cases) {
      test(c['id'], () {
        final messages = [
          for (final raw
              in (c['messages'] as List).cast<Map<String, dynamic>>())
            message(raw),
        ];
        final items = flareImageGalleryItems(messages);
        expect([
          for (final item in items) '${item.messageId}#${item.index}',
        ], c['items']);
        for (final open in (c['opens'] as List).cast<Map<String, dynamic>>()) {
          expect(
            flareImageGalleryStart(
              items,
              open['message'] as String,
              open['index'] as int,
            ),
            open['start'],
            reason: '${open['message']}#${open['index']}',
          );
        }
      });
    }
  });

  group('gallery preview', () {
    final images = [
      for (var i = 0; i < 3; i++)
        FlareImageContent(url: 'https://cdn.example/$i.jpg'),
    ];

    testWidgets('pages with its side keys and says where it is', (
      tester,
    ) async {
      await tester.pumpWidget(
        MaterialApp(
          home: Builder(
            builder: (context) => TextButton(
              onPressed: () => FlareImagePreview.presentGallery(
                context,
                images: images,
                initialIndex: 1,
              ),
              child: const Text('open'),
            ),
          ),
        ),
      );
      await tester.tap(find.text('open'));
      await tester.pumpAndSettle();
      expect(find.text('2 / 3'), findsOneWidget);
      expect(find.bySemanticsLabel('第 2 张，共 3 张'), findsOneWidget);
      await tester.tap(find.bySemanticsLabel('下一张'));
      await tester.pump();
      expect(find.text('3 / 3'), findsOneWidget);
      // At the last image the next key does nothing.
      await tester.tap(find.bySemanticsLabel('下一张'));
      await tester.pump();
      expect(find.text('3 / 3'), findsOneWidget);
      await tester.tap(find.bySemanticsLabel('上一张'));
      await tester.tap(find.bySemanticsLabel('上一张'));
      await tester.pump();
      expect(find.text('1 / 3'), findsOneWidget);
    });

    testWidgets('a sideways swipe pages', (tester) async {
      await tester.pumpWidget(
        MaterialApp(
          home: Builder(
            builder: (context) => TextButton(
              onPressed: () =>
                  FlareImagePreview.presentGallery(context, images: images),
              child: const Text('open'),
            ),
          ),
        ),
      );
      await tester.tap(find.text('open'));
      await tester.pumpAndSettle();
      expect(find.text('1 / 3'), findsOneWidget);
      await tester.fling(
        find.byType(InteractiveViewer),
        const Offset(-300, 0),
        1500,
      );
      await tester.pumpAndSettle();
      expect(find.text('2 / 3'), findsOneWidget);
    });

    testWidgets('a single image shows no paging chrome', (tester) async {
      await tester.pumpWidget(
        const MaterialApp(
          home: FlareImagePreview(
            show: true,
            imageSrc: 'https://cdn.example/a.jpg',
          ),
        ),
      );
      expect(find.bySemanticsLabel('下一张'), findsNothing);
      expect(find.textContaining(' / '), findsNothing);
    });

    testWidgets('a picture outside a timeline offers the download its body '
        'was given, and none without one', (tester) async {
      final downloads = <String>[];
      final picture = FlareImageContent(
        url: 'https://cdn.example/solo.jpg',
        alt: '海边',
      );
      await tester.pumpWidget(
        MaterialApp(
          home: Scaffold(
            body: Column(
              children: [
                FlareMessageContentView(
                  content: picture,
                  onMediaDownload: (content) =>
                      downloads.add((content as FlareImageContent).url),
                ),
              ],
            ),
          ),
        ),
      );
      await tester.tap(find.bySemanticsLabel('海边'));
      await tester.pumpAndSettle();
      await tester.tap(find.bySemanticsLabel('下载'));
      await tester.pump();
      expect(downloads, ['https://cdn.example/solo.jpg']);

      // A new app, not the same one rebuilt: the first preview's route must not survive.
      await tester.pumpWidget(
        MaterialApp(
          key: UniqueKey(),
          home: Scaffold(
            body: Column(children: [FlareMessageContentView(content: picture)]),
          ),
        ),
      );
      await tester.pumpAndSettle();
      await tester.tap(find.bySemanticsLabel('海边'));
      await tester.pumpAndSettle();
      expect(find.bySemanticsLabel('下载'), findsNothing);
    });

    testWidgets('the timeline gallery saves the picture on screen with the '
        'message it belongs to', (tester) async {
      final downloads = <String>[];
      final messages = [
        FlareMessageData(
          id: 'm1',
          senderId: 'u',
          senderName: 'U',
          content: images[0],
        ),
        FlareMessageData(
          id: 'm2',
          senderId: 'u',
          senderName: 'U',
          content: FlareImageGroupContent(images: [images[1], images[2]]),
        ),
      ];
      await tester.pumpWidget(
        MaterialApp(
          home: Scaffold(
            body: FlareMessageList(
              messages: messages,
              currentUserId: 'me',
              onMediaDownload: (message, content) => downloads.add(
                '${message.id}:${(content as FlareImageContent).url}',
              ),
            ),
          ),
        ),
      );
      await tester.pumpAndSettle();
      await tester.tap(find.bySemanticsLabel(RegExp('第 2 张图片')));
      await tester.pumpAndSettle();
      expect(find.text('3 / 3'), findsOneWidget);
      await tester.tap(find.bySemanticsLabel('下载'));
      await tester.tap(find.bySemanticsLabel('上一张'));
      await tester.tap(find.bySemanticsLabel('上一张'));
      await tester.pump();
      expect(find.text('1 / 3'), findsOneWidget);
      await tester.tap(find.bySemanticsLabel('下载'));
      await tester.pump();
      expect(downloads, [
        'm2:https://cdn.example/2.jpg',
        'm1:https://cdn.example/0.jpg',
      ]);
    });

    testWidgets('an image in a timeline opens the timeline\'s gallery at it', (
      tester,
    ) async {
      final messages = [
        FlareMessageData(
          id: 'm1',
          senderId: 'u',
          senderName: 'U',
          content: images[0],
        ),
        FlareMessageData(
          id: 'm2',
          senderId: 'u',
          senderName: 'U',
          content: FlareImageGroupContent(images: [images[1], images[2]]),
        ),
      ];
      await tester.pumpWidget(
        MaterialApp(
          home: Scaffold(
            body: FlareMessageList(messages: messages, currentUserId: 'me'),
          ),
        ),
      );
      await tester.pumpAndSettle();
      await tester.tap(find.bySemanticsLabel(RegExp('第 2 张图片')));
      await tester.pumpAndSettle();
      expect(find.text('3 / 3'), findsOneWidget);
    });
  });
}
